File: tools/ops-eye/capture.py
#!/usr/bin/env python3
"""Multimodal Ops Eye capture tool.

Captures macOS screenshots, runs local OCR with tesseract, and returns structured JSON.
"""

from __future__ import annotations

import argparse
import json
import re
import shutil
import subprocess
import sys
import tempfile
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

ERROR_PATTERNS = [
    r"\berror\b",
    r"\bfailed\b",
    r"\bexception\b",
    r"\btraceback\b",
    r"\bwarning\b",
    r"\bdenied\b",
    r"\bunavailable\b",
    r"\bcrash(ed|ing)?\b",
]


def run(cmd: list[str]) -> subprocess.CompletedProcess[str]:
    return subprocess.run(cmd, text=True, capture_output=True)


def resolve_binary(name: str, fallbacks: list[str] | None = None) -> str | None:
    found = shutil.which(name)
    if found:
        return found
    for p in fallbacks or []:
        if Path(p).exists():
            return p
    return None


def ensure_dependencies() -> tuple[str, str]:
    screencapture_bin = resolve_binary("screencapture", ["/usr/sbin/screencapture", "/usr/bin/screencapture"])
    tesseract_bin = resolve_binary("tesseract", ["/opt/homebrew/bin/tesseract", "/usr/local/bin/tesseract"])

    if screencapture_bin is None:
        raise RuntimeError("screencapture is not available on this system")
    if tesseract_bin is None:
        raise RuntimeError("tesseract is not installed. Try: brew install tesseract")

    return screencapture_bin, tesseract_bin


def capture_screenshot(screencapture_bin: str, path: Path, mode: str, window_id: str | None) -> dict[str, Any]:
    cmd = [screencapture_bin, "-x"]

    if mode == "window":
        if not window_id:
            raise RuntimeError("--window-id is required when --mode window")
        cmd.extend(["-l", str(window_id)])

    cmd.append(str(path))
    proc = run(cmd)
    if proc.returncode != 0:
        raise RuntimeError(f"screencapture failed: {proc.stderr.strip() or proc.stdout.strip()}")

    return {
        "path": str(path),
        "bytes": path.stat().st_size if path.exists() else 0,
        "mode": mode,
        "window_id": window_id,
    }


def ocr_image(tesseract_bin: str, image_path: Path, lang: str = "eng") -> dict[str, Any]:
    cmd = [tesseract_bin, str(image_path), "stdout", "-l", lang, "--psm", "6"]
    proc = run(cmd)
    if proc.returncode != 0:
        raise RuntimeError(f"tesseract failed: {proc.stderr.strip() or proc.stdout.strip()}")

    text = proc.stdout or ""
    cleaned = text.strip()
    lines = [ln for ln in cleaned.splitlines() if ln.strip()]

    return {
        "text": cleaned,
        "line_count": len(lines),
        "char_count": len(cleaned),
        "language": lang,
    }


def get_frontmost_app() -> dict[str, str | None]:
    script = (
        'tell application "System Events"\n'
        'set p to first process whose frontmost is true\n'
        'set appName to name of p\n'
        'set winName to ""\n'
        'try\n'
        'set winName to name of front window of p\n'
        'end try\n'
        'return appName & "|||" & winName\n'
        'end tell'
    )

    proc = run(["osascript", "-e", script])
    if proc.returncode != 0:
        return {"app_name": None, "window_title": None}

    out = (proc.stdout or "").strip()
    if "|||" in out:
        app, win = out.split("|||", 1)
        return {"app_name": app or None, "window_title": win or None}
    return {"app_name": out or None, "window_title": None}


def detect_ui_state(ocr_text: str, frontmost: dict[str, Any]) -> dict[str, Any]:
    lowered = (ocr_text or "").lower()
    matches = []
    for pattern in ERROR_PATTERNS:
        if re.search(pattern, lowered, re.IGNORECASE):
            matches.append(pattern)

    signals = []
    if matches:
        signals.append("error_keywords_detected")

    title = (frontmost.get("window_title") or "").lower()
    app_name = (frontmost.get("app_name") or "").lower()
    if "dialog" in title or "alert" in title:
        signals.append("dialog_window_title")
    if any(k in title for k in ["error", "failed", "warning"]) or any(
        k in app_name for k in ["crash", "report", "installer"]
    ):
        signals.append("possible_error_window")

    severity = "normal"
    if "possible_error_window" in signals or matches:
        severity = "warning"

    return {
        "severity": severity,
        "signals": signals,
        "error_pattern_matches": matches,
    }


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Capture screenshot + OCR + UI state JSON")
    parser.add_argument("--mode", choices=["full", "window"], default="full")
    parser.add_argument("--window-id", help="CGWindowID for --mode window")
    parser.add_argument("--output-image", help="Save screenshot to this path")
    parser.add_argument("--lang", default="eng", help="Tesseract language (default: eng)")
    parser.add_argument(
        "--ui-state",
        action="store_true",
        help="Include UI state detection (frontmost app + error heuristics)",
    )
    parser.add_argument(
        "--keep-temp",
        action="store_true",
        help="Keep temp screenshot when --output-image is not provided",
    )
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    started = datetime.now(timezone.utc)

    try:
        screencapture_bin, tesseract_bin = ensure_dependencies()

        explicit_image = Path(args.output_image).expanduser() if args.output_image else None
        if explicit_image:
            explicit_image.parent.mkdir(parents=True, exist_ok=True)
            image_path = explicit_image
            temp_used = False
        else:
            tmp_dir = Path(tempfile.mkdtemp(prefix="ops-eye-"))
            image_path = tmp_dir / "capture.png"
            temp_used = True

        capture_meta = capture_screenshot(screencapture_bin, image_path, args.mode, args.window_id)
        ocr_meta = ocr_image(tesseract_bin, image_path, lang=args.lang)

        frontmost = get_frontmost_app() if args.ui_state else {}
        ui_state = detect_ui_state(ocr_meta["text"], frontmost) if args.ui_state else None

        payload: dict[str, Any] = {
            "ok": True,
            "timestamp_utc": started.isoformat(),
            "capture": capture_meta,
            "ocr": ocr_meta,
            "frontmost": frontmost if args.ui_state else None,
            "ui_state": ui_state,
            "engine": {"ocr": "tesseract", "platform": "macOS"},
        }

        if temp_used and not args.keep_temp:
            try:
                image_path.unlink(missing_ok=True)
                image_path.parent.rmdir()
                payload["capture"]["path"] = None
            except Exception:
                pass

        print(json.dumps(payload, indent=2))
        return 0

    except Exception as exc:
        err = {
            "ok": False,
            "error": str(exc),
            "timestamp_utc": started.isoformat(),
        }
        print(json.dumps(err, indent=2), file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/ops-eye/capture.ts
#!/usr/bin/env npx tsx

import { spawnSync } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import os from "node:os";

const ERROR_PATTERNS = [
  "\\berror\\b",
  "\\bfailed\\b",
  "\\bexception\\b",
  "\\btraceback\\b",
  "\\bwarning\\b",
  "\\bdenied\\b",
  "\\bunavailable\\b",
  "\\bcrash(ed|ing)?\\b",
];

function run(cmd: string[]): { code: number; stdout: string; stderr: string } {
  const proc = spawnSync(cmd[0], cmd.slice(1), { encoding: "utf8" });
  return {
    code: proc.status ?? 1,
    stdout: proc.stdout || "",
    stderr: proc.stderr || "",
  };
}

function resolveBinary(name: string, fallbacks: string[] = []): string | null {
  const which = spawnSync("which", [name], { encoding: "utf8" });
  const found = (which.stdout || "").trim();
  if (which.status === 0 && found) return found;
  for (const p of fallbacks) {
    if (fs.existsSync(p)) return p;
  }
  return null;
}

function ensureDependencies(): [string, string] {
  const screencaptureBin = resolveBinary("screencapture", ["/usr/sbin/screencapture", "/usr/bin/screencapture"]);
  const tesseractBin = resolveBinary("tesseract", ["/opt/homebrew/bin/tesseract", "/usr/local/bin/tesseract"]);
  if (!screencaptureBin) throw new Error("screencapture is not available on this system");
  if (!tesseractBin) throw new Error("tesseract is not installed. Try: brew install tesseract");
  return [screencaptureBin, tesseractBin];
}

function captureScreenshot(
  screencaptureBin: string,
  imagePath: string,
  mode: string,
  windowId: string | null,
): Record<string, any> {
  const cmd = [screencaptureBin, "-x"];
  if (mode === "window") {
    if (!windowId) throw new Error("--window-id is required when --mode window");
    cmd.push("-l", String(windowId));
  }
  cmd.push(imagePath);
  const proc = run(cmd);
  if (proc.code !== 0) {
    throw new Error(`screencapture failed: ${(proc.stderr || proc.stdout).trim()}`);
  }
  const bytes = fs.existsSync(imagePath) ? fs.statSync(imagePath).size : 0;
  return { path: imagePath, bytes, mode, window_id: windowId };
}

function ocrImage(tesseractBin: string, imagePath: string, lang = "eng"): Record<string, any> {
  const cmd = [tesseractBin, imagePath, "stdout", "-l", lang, "--psm", "6"];
  const proc = run(cmd);
  if (proc.code !== 0) {
    throw new Error(`tesseract failed: ${(proc.stderr || proc.stdout).trim()}`);
  }
  const text = proc.stdout || "";
  const cleaned = text.trim();
  const lines = cleaned.split(/\r?\n/).filter((ln) => ln.trim());
  return {
    text: cleaned,
    line_count: lines.length,
    char_count: cleaned.length,
    language: lang,
  };
}

function getFrontmostApp(): Record<string, string | null> {
  const script =
    'tell application "System Events"\n' +
    'set p to first process whose frontmost is true\n' +
    'set appName to name of p\n' +
    'set winName to ""\n' +
    'try\n' +
    'set winName to name of front window of p\n' +
    'end try\n' +
    'return appName & "|||" & winName\n' +
    'end tell';

  const proc = run(["osascript", "-e", script]);
  if (proc.code !== 0) return { app_name: null, window_title: null };
  const out = (proc.stdout || "").trim();
  if (out.includes("|||")) {
    const [app, win] = out.split("|||", 2);
    return { app_name: app || null, window_title: win || null };
  }
  return { app_name: out || null, window_title: null };
}

function detectUiState(ocrText: string, frontmost: Record<string, any>): Record<string, any> {
  const lowered = (ocrText || "").toLowerCase();
  const matches: string[] = [];
  for (const pattern of ERROR_PATTERNS) {
    if (new RegExp(pattern, "i").test(lowered)) matches.push(pattern);
  }

  const signals: string[] = [];
  if (matches.length) signals.push("error_keywords_detected");

  const title = String(frontmost.window_title ?? "").toLowerCase();
  const appName = String(frontmost.app_name ?? "").toLowerCase();
  if (title.includes("dialog") || title.includes("alert")) signals.push("dialog_window_title");
  if (["error", "failed", "warning"].some((k) => title.includes(k)) ||
      ["crash", "report", "installer"].some((k) => appName.includes(k))) {
    signals.push("possible_error_window");
  }

  let severity = "normal";
  if (signals.includes("possible_error_window") || matches.length) severity = "warning";

  return { severity, signals, error_pattern_matches: matches };
}

function parseArgs(argv: string[]) {
  const args = {
    mode: "full",
    windowId: null as string | null,
    outputImage: null as string | null,
    lang: "eng",
    uiState: false,
    keepTemp: false,
  };

  for (let i = 0; i < argv.length; i += 1) {
    const a = argv[i];
    if (a === "--mode") args.mode = argv[++i] ?? args.mode;
    else if (a === "--window-id") args.windowId = argv[++i] ?? null;
    else if (a === "--output-image") args.outputImage = argv[++i] ?? null;
    else if (a === "--lang") args.lang = argv[++i] ?? args.lang;
    else if (a === "--ui-state") args.uiState = true;
    else if (a === "--keep-temp") args.keepTemp = true;
  }

  return args;
}

async function main(): Promise<number> {
  const args = parseArgs(process.argv.slice(2));
  const started = new Date().toISOString();

  try {
    const [screencaptureBin, tesseractBin] = ensureDependencies();

    let imagePath: string;
    let tempUsed = false;

    if (args.outputImage) {
      const explicit = path.resolve(args.outputImage);
      fs.mkdirSync(path.dirname(explicit), { recursive: true });
      imagePath = explicit;
    } else {
      const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "ops-eye-"));
      imagePath = path.join(tmpDir, "capture.png");
      tempUsed = true;
    }

    const captureMeta = captureScreenshot(screencaptureBin, imagePath, args.mode, args.windowId);
    const ocrMeta = ocrImage(tesseractBin, imagePath, args.lang);

    const frontmost = args.uiState ? getFrontmostApp() : {};
    const uiState = args.uiState ? detectUiState(ocrMeta.text, frontmost) : null;

    const payload: Record<string, any> = {
      ok: true,
      timestamp_utc: started,
      capture: captureMeta,
      ocr: ocrMeta,
      frontmost: args.uiState ? frontmost : null,
      ui_state: uiState,
      engine: { ocr: "tesseract", platform: "macOS" },
    };

    if (tempUsed && !args.keepTemp) {
      try {
        fs.unlinkSync(imagePath);
        fs.rmdirSync(path.dirname(imagePath));
        payload.capture.path = null;
      } catch {
        // ignore
      }
    }

    console.log(JSON.stringify(payload, null, 2));
    return 0;
  } catch (err) {
    const payload = {
      ok: false,
      error: err instanceof Error ? err.message : String(err),
      timestamp_utc: started,
    };
    console.error(JSON.stringify(payload, null, 2));
    return 1;
  }
}

main()
  .then((code) => process.exit(code))
  .catch((err) => {
    console.error(err instanceof Error ? err.message : String(err));
    process.exit(1);
  });
